pkg/controller/daemon/daemoncontroller.go

```diff
@@ -273,6 +289,138 @@ func (dsc *DaemonSetsController) getPodDaemonSets(pod *v1.Pod) []*extensions.Dae
 	return sets
 }
 
+// getDaemonSetsForHistory returns a list of DaemonSets that potentially
+// match a ControllerRevision.
+func (dsc *DaemonSetsController) getDaemonSetsForHistory(history *apps.ControllerRevision) []*extensions.DaemonSet {
+	daemonSets, err := dsc.dsLister.GetHistoryDaemonSets(history)
+	if err != nil || len(daemonSets) == 0 {
+		return nil
+	}
+	if len(daemonSets) > 1 {
+		// ControllerRef will ensure we don't do anything crazy, but more than one
+		// item in this list nevertheless constitutes user error.
+		glog.V(4).Infof("User error! more than one DaemonSets is selecting ControllerRevision %s/%s with labels: %#v",
+			history.Namespace, history.Name, history.Labels)
+	}
+	return daemonSets
+}
+
+// addHistory enqueues the DaemonSet that manages a ControllerRevision when the ControllerRevision is created.
+func (dsc *DaemonSetsController) addHistory(obj interface{}) {
+	history := obj.(*apps.ControllerRevision)
+	if history.DeletionTimestamp != nil {
+		// On a restart of the controller manager, it's possible for an object to
+		// show up in a state that is already pending deletion.
+		dsc.deleteHistory(history)
+		return
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if controllerRef := controller.GetControllerOf(history); controllerRef != nil {
+		ds := dsc.resolveControllerRef(history.Namespace, controllerRef)
+		if ds == nil {
+			return
+		}
+		glog.V(4).Infof("ControllerRevision %s added.", history.Name)
+		dsc.enqueueDaemonSet(ds)
+		return
+	}
+
+	// Otherwise, it's an orphan. Get a list of all matching DaemonSets and sync
+	// them to see if anyone wants to adopt it.
+	daemonSets := dsc.getDaemonSetsForHistory(history)
+	if len(daemonSets) == 0 {
+		return
+	}
+	glog.V(4).Infof("Orphan ControllerRevision %s added.", history.Name)
+	for _, ds := range daemonSets {
+		dsc.enqueueDaemonSet(ds)
+	}
+}
+
+// updateHistory figures out what DaemonSet(s) manage a ControllerRevision when the ControllerRevision
+// is updated and wake them up. If the anything of the ControllerRevision have changed, we need to
```

Comment:
From the proposal, only its `.data` cannot be updated. We need to at least be able to update its owner ref, otherwise we can't handle orphan / adoption / garbage collection.